openhtf/util/data.py

```diff
@@ -135,21 +136,27 @@ def convert_to_base_types(obj, ignore_keys=tuple(), tuple_type=tuple):
 
   # Recursively convert values in dicts, lists, and tuples.
   if isinstance(obj, dict):
-    obj = {convert_to_base_types(k, ignore_keys, tuple_type):
+    return {convert_to_base_types(k, ignore_keys, tuple_type):
                convert_to_base_types(v, ignore_keys, tuple_type)
-           for k, v in obj.iteritems() if k not in ignore_keys}
+            for k, v in obj.iteritems() if k not in ignore_keys}
   elif isinstance(obj, list):
-    obj = [convert_to_base_types(val, ignore_keys, tuple_type) for val in obj]
+    return [convert_to_base_types(val, ignore_keys, tuple_type) for val in obj]
   elif isinstance(obj, tuple):
-    obj = tuple_type(
+    return tuple_type(
         convert_to_base_types(value, ignore_keys, tuple_type) for value in obj)
-  elif obj is not None and (
-      not isinstance(obj, numbers.Number) and not isinstance(obj, basestring)):
-    # Leave None as None to distinguish it from "None", as well as numbers and
-    # strings, converting anything unknown to strings.
-    obj = str(obj)
 
-  return obj
+  # Preserve booleans, None, and unicode.
+  elif obj in (True, False, None) or type(obj) is unicode:
```

Comment:
maybe include int and long as well, and do this check first since it's more likely (more leaves than parents in the tree)